torchao/csrc/metal/test_int4_q_kernels.mm

```diff
@@ -0,0 +1,291 @@
+        newFunctionWithName:[NSString
+                                stringWithFormat:@"int4pack_mv_%u_%s",
+                                                 groupSize,
+                                                 type_string<T>().c_str()]];
+    if (func == nil) {
+      fail("Can:t get function");
+    }
+    NSError *error = nil;
+    auto cpl = [lib.device newComputePipelineStateWithFunction:func
+                                                         error:&error];
+    if (cpl == nil) {
+      fail("Failed to construct pipeline state: ",
+           error.description.UTF8String);
+    }
+    id<MTLCommandQueue> queue = [lib.device newCommandQueue];
+    auto do_compute = ^() {
+      @autoreleasepool {
+        auto desc = [MTLCommandBufferDescriptor new];
+        desc.errorOptions = MTLCommandBufferErrorOptionEncoderExecutionStatus;
+        id<MTLCommandBuffer> cmdBuffer =
+            [queue commandBufferWithDescriptor:desc];
+        encodeMM(cmdBuffer, cpl);
+        [cmdBuffer commit];
+        [cmdBuffer waitUntilCompleted];
+      }
+    };
+
+    do_compute();
+
+    if (!validate<T>()) {
+      return false;
+    }
+    return true;
+  }
+
+private:
+  template <typename T> std::string type_string() const;
+  template <> std::string type_string<BFloat16>() const { return "bfloat"; }
+  template <> std::string type_string<float>() const { return "float"; }
+  template <> std::string type_string<Float16>() const { return "half"; }
+  void allocBuffers(id<MTLDevice> device, const unsigned elem_size = 4) {
+    buf_A = allocSharedBuffer(device, M * K * elem_size);
+    buf_B = allocSharedBuffer(device, N * K / 2);
+    buf_C = allocSharedBuffer(device, M * N * elem_size);
+    buf_SZ = allocSharedBuffer(device, N * K / groupSize * 2 * elem_size);
+  }
+
+public:
+  unsigned M, N, K;     // Input-output matirx dims
+  id<MTLBuffer> buf_A;  // MxK elements
+  id<MTLBuffer> buf_B;  // NxK elements
+  id<MTLBuffer> buf_C;  // MxN elements
+  id<MTLBuffer> buf_SZ; // (K/groupSize)xNx2 elements
+  id<MTLLibrary> lib;
+  std::string lib_name;
+};
+
+template <unsigned groupSize> struct Int4MV : public Int4MMBase<groupSize> {
+  using Int4MMBase<groupSize>::M;
+  using Int4MMBase<groupSize>::N;
+  Int4MV(id<MTLDevice> device, const std::string &lib_name_, unsigned M_,
+         unsigned N_, unsigned K_)
+      : Int4MMBase<groupSize>(device, lib_name_, M_, N_, K_) {
+    if (M != 1) {
+      fail("Value of M must be 1");
+    }
+  }
+  void dispatchThreads(id<MTLComputeCommandEncoder> encoder,
+                       unsigned maxThreadsPerGroup) const override {
+    constexpr auto blockSize = 8;
+    if (maxThreadsPerGroup < blockSize * blockSize) {
+      throw std::runtime_error("Can't dispatch!");
+    }
+    [encoder dispatchThreads:MTLSizeMake(N / 4 * 32, 1, M)
+        threadsPerThreadgroup:MTLSizeMake(64, 1, 1)];
+  }
+};
+
+int main() {
+  unsigned M, N, K;
+  std::tie(M, N, K) = std::make_tuple(1, 4096, 4096);
```

Comment:
It looks like MVP, could we test with M > 1?

Reply:
THis is WIP. more stuff will land to test more stuff. The first version hasnt even tested beyond groupd size 32 and I dont intend it to be so. So will definitely be updating all this in subsequent diffs